### ISSUE TITLE
Specify how to enable an event stream

### ISSUE THREAD

**Issue by jgraham** [w3c/webdriver-bidi#17]
For various use cases (e.g. logging) we want the ability to enable a stream of events from the browser to the client. But it isn't possible to have all events enabled all the time; we want a design that allows us to enable and disable particular events, either for the session as a whole or for specific browsing contexts. In some cases enabling a specific event stream might also enable additional commands that can make use of data from the events.

The explainer document currently proposes a mechanism where different event streams have enable/disable pairs with a counter that is incremented for each enable and decremented for each disable (clamped at 0), with the stream only being actually disabled when the count reaches 0. But there are concerns that this mechanism may be over-complex, providing a protocol-level solution to application-level concern around how to handle multiple parts of the code. 

The simplest solution here seems to be to divide events into various groups ("domains" in CDP) and have an enable and disable command for each group. Some additional property may be needed to determine the scope of the enabling (e.g. one browsing context / agent vs global).

**Comment by bwalderman** [comment]
Here's a strawman proposal for enabling events:

Related events are grouped into domains. The client can subscribe to all events in a domain, or to specific events in a domain. Events can be enabled globally for all contexts, or for a subset of contexts. When events are enabled globally, those events are received from all current and future contexts.

Most events are disabled initially. When a new WebDriver BiDi session is created, the client should enable any event streams they are interested in before proceeding with their test. It might be a good idea to enable events related to context discovery by default; Something like CDP's `Target` domain.

Subscribing to events globally is useful for automation scenarios where you want to be aware of everything that is happening throughout the browser. Sometimes though, you may want to enable events for a particular tab, so we should make it possible to specify a top-level browsing context or service worker to enable events on. Enabling events for a particular nested frame or a particular worker doesn't seem as useful so we probably don't need to make this a priority. 

If we follow CDP's general approach, and add parameters for choosing specific events and contexts we might end up with something like this:

```js
// Enable these specific Network events globally
{
    method: "Network.enable",
    events: ["requestWillBeSent", "responseReceived"] 
}

// Enable all Network.* events for this specific context
{
    method: "Network.enable",
    contexts: ["<tab A>", "<service worker B>"]
}
```

The values in the `contexts` array must be top-level contexts: a tab, service worker, or shared worker. The requested events will be enabled for that context, and all nested contexts such as frames and workers.

It might also be useful to allow the client to subscribe to events for particular contexts even if they don't know those context's IDs yet. For example, to subscribe to event for any current or future tab matching a certain URL:

```js
{
    method: "Network.enable",
    contexts: [{ url: "https://example.com/*" }]
}
```


*Note*: I recently proposed grouping events by sender instead of by domain, and adding `to`/`from` property to messages to subscribe to events and indicate the sender: https://github.com/w3c/webdriver-bidi/pull/41. However, this requires the client to know a context's ID before they can subscribe to any events from it, so I've decided to abandon that idea.

**Comment by jgraham** [comment]
I like where this is going.

Some questions:

Why not just make the method `Enable` or `Protocol.enable` or something, if we want everything to have a two-part name. Then in the events array we allow dotted names:

```
{
  method: "Protocol.Enable",
  events: ["Network.requestWillBeSent", "Network.responseRecieved", "Logging"]
}
```

In this case we would get the two events from the networking domain and everything from logging.

Where does the restriction on enabling only for top-level browsing contexts come from? Does that match CDP? Intuitively it's not obvious why you would be unable to get e.g. network events just for an iframe or something.

I propose we also have a `Protocol.disable` command with the opposite semantics.

(aside: I assume the syntax here is a placeholder for the real thing with everything apart from the name in a params field).

**Comment by bwalderman** [comment]
I like the suggestion for a single general-purpose `Enable` command. Also +1 on a `disable` command with opposite semantics. For clarity, I'd prefer to call these commands `subscribe` and `unsubscribe`. 

I proposed limiting subscriptions to top-level contexts because things may get tricky if we allow subscriptions for nested context. Some of this depends on the object model we decide on and what kinds of contexts are addressable. For the moment, let's assume we have a model where browsing contexts and their associated JS realm are different contexts with their own identifiers (a subject for discussion in #18).

Different events, especially events in different domains, may come from different kinds of contexts. `Network` events would likely come from a browsing context, whereas something like `Runtime.exceptionThrown` would come from a JS realm. To subscribe to events that come from different types of contexts, we would need a way to specify the contexts per-event. For example:

```js
{
    method: "subscribe",
    params: {
        subscriptions: [
            {
                events: ["Network.requestWillBeSent", "Network.responseReceived"],
                contexts: ["<a nested browsing context>", "<another nested browsing context>"]
            },
            {
                events: ["Logging"],
                // Everywhere
            },
            {
                events: ["Runtime.exceptionThrown"],
                contexts: ["<a js realm>"]
            }
        ]
    }
}
```

This would allow you to filter for events from specific contexts you are interested in but doesn't consider subtrees. For example, if you are interested in the network activity coming from a particular iframe, chances are you are also interested in activity from its nested iframes as well. So in addition to subscribing to events globally, or for a specific context, we'd also want a way to say "subscribe to this context, and its subtree".

I'm not sure the subscription API needs to be quite this fine-grained, so limiting subscriptions to top-level browsing contexts seems like a fair compromise. With this approach, clients can subscribe to events globally, or limit events to top-level contexts of interest and their subtree. I think this would address the most common use cases.

**Comment by jgraham** [comment]
Fixed by #51
